phpunit/GutenbergTestSuiteLoader.php

```diff
@@ -0,0 +1,119 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace phpunit;
+
+use PHPUnit\Framework\TestCase;
+use PHPUnit\Util\FileLoader;
+use PHPUnit\Runner\TestSuiteLoader;
+
+use \Exception;
+use \ReflectionClass;
+use \ReflectionException;
+
+/**
+ * @internal This class is not covered by the backward compatibility promise for PHPUnit
+ *
+ * @deprecated see https://github.com/sebastianbergmann/phpunit/issues/4039
+ */
+class GutenbergTestSuiteLoader implements TestSuiteLoader
+{
+	/**
+	 * @throws Exception
+	 */
+	public function load(string $suiteClassFile): ReflectionClass
+	{
+		$suiteClassName = basename($suiteClassFile, '.php');
+		$loadedClasses  = get_declared_classes();
+
+		if (!class_exists($suiteClassName, false)) {
+			/* @noinspection UnusedFunctionResultInspection */
+			FileLoader::checkAndLoad($suiteClassFile);
+
+			$loadedClasses = array_values(
+				array_diff(get_declared_classes(), $loadedClasses)
```

Comment:
it's confusing to read this because it feels circular. also, it reads as if this should be an empty array since we're loading the same file that is coming in from this call into `load()`

at a minimum could we document why it's different now after `checkAndLoad` on the given filename?

it could help for comprehensibility to create a new intermediate variable with a different name.

```php
/*
 * If we didn't find the class it could be a case-folding issue.
 * For phpunit through version 8 it would find the appropriate
 * class to use, but at version 9 it started requiring that the
 * name of the class exactly match the name of the file.
 *
 * Here we're going to get a list of all the classes defined in
 * the test file so we can search through them to see if we have
 * a lexical variation of the class name that matches the filename.
 */

FileLoader::checkAndLoad( $suiteClassFile );
$allClassesFromLoadedFile = get_declared_classes();

if ( empty( $allClassesFromLoadedFiles ) ) {
	throw …
}

$loadedClasses = $allClassesFromLoadedFile;
```

Reply:
Oh that's just copy and paste from the PHPUnit project. I'd inherit from their `StandardTestSuiteLoader` class, but unfortunately it's marked as `final`. I'd rather avoid refactoring the copy-pasta, too, because we might need to pull in some updates in the future. I'm noodling on a clearer distinction between the original code and our customization.

Reply:
Actually I'm starting to think that ditching the original code and sticking with a fully custom loader could be actually a better way to go – perhaps it would provide consistent loading mechanics across the future PHPUnit versions.